scanner/src/main/kotlin/scanners/fossid/FossId.kt

```diff
@@ -0,0 +1,396 @@
+import org.ossreviewtoolkit.clients.fossid.model.status.DownloadStatus
+import org.ossreviewtoolkit.clients.fossid.model.status.ScanState
+import org.ossreviewtoolkit.clients.fossid.runScan
+import org.ossreviewtoolkit.clients.fossid.toList
+import org.ossreviewtoolkit.model.Package
+import org.ossreviewtoolkit.model.Provenance
+import org.ossreviewtoolkit.model.ScanResult
+import org.ossreviewtoolkit.model.ScanSummary
+import org.ossreviewtoolkit.model.VcsType
+import org.ossreviewtoolkit.model.config.ScannerConfiguration
+import org.ossreviewtoolkit.scanner.AbstractScannerFactory
+import org.ossreviewtoolkit.scanner.RemoteScanner
+import org.ossreviewtoolkit.utils.log
+
+import retrofit2.Retrofit
+import retrofit2.converter.jackson.JacksonConverterFactory
+
+/**
+ * A wrapper for [FossID](https://fossid.com/).
+ *
+ * This scanner can be configured in [ScannerConfiguration.options] using the key "FossId". It offers the following
+ * configuration options:
+ *
+ * * **"serverUrl":** The URL of the FossID server.
+ * * **"user":** The user to connect to the FossID server.
+ * * **"apiKey":** The API key of the user which connects to the FossID server.
+ */
+class FossId(name: String, config: ScannerConfiguration) : RemoteScanner(name, config) {
+    class Factory : AbstractScannerFactory<FossId>("FossId") {
+        override fun create(config: ScannerConfiguration) = FossId(scannerName, config)
+    }
+
+    companion object {
+        @JvmStatic
+        private val PROJECT_NAME_PATTERN = Pattern.compile("""^.*\/([\w-]+)\.git$""")
+
+        @JvmStatic
+        private val GIT_FETCH_DONE_PATTERN = Pattern.compile("-> FETCH_HEAD")
+
+        @JvmStatic
+        private val SECRET = "X".repeat(6)
+
+        @JvmStatic
+        private val WAIT_INTERVAL_MS = 10000L
+
+        @JvmStatic
+        private val WAIT_REPETITION = 50
+
+        /**
+         * Convert a GIT repo URL to a valid project name :
+         * https://github.com/jshttp/mime-types.git -> mime-types
+         */
+        fun convertGitUrlToProjectName(gitRepoUrl: String): String {
+            val projectNameMatcher = PROJECT_NAME_PATTERN.matcher(gitRepoUrl)
+
+            require(projectNameMatcher.matches()) {
+                "GIT repo url '$gitRepoUrl' cannot be matched"
+            }
+
+            val projectCode = projectNameMatcher.group(1)
+            log.info { "Project code for url='$gitRepoUrl is '$projectCode'." }
+            return projectCode
+        }
+    }
+
+    private val serverUrl: String
+    private val apiKey: String
+    private val user: String
+    private val service: FossIdRestService
+
+    // TODO parse the version from the FossID login page
+    override val version = ""
+
+    override val configuration: String
+        get() {
+            // sanitized configuration to hide sensitive information
+            return "--serverUrl=$SECRET --apiKey=$SECRET --user=$SECRET"
```

Comment:
> More generally, we could discuss whether having a configuration makes sense for remote scanners at all. If we'd never store scan results from remote scanners but always rely on their internal caching, then we'd probably also never need this `configuration` field.

I would not rule out keeping the configuration for remote scanners.
According to companies policies, it can be useful to keep a trace of remote server URLs, API user or even credentials (e.g. if one time token or password was used).

However for FossID, I will simply blank the configuration as suggested.